src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs

```diff
@@ -1306,6 +1310,32 @@ internal class PipelineStopper
         /// Object used for synchronization.
         /// </summary>
         private object _syncRoot = new object();
+
+        internal void AwaitPipelineCriticalSection()
+        {
+            _criticalSectionSemaphore.Wait();
+            _criticalSectionSemaphore.Release();
+        }
+
+        internal void EnterCriticalSection()
+        {
+            Interlocked.Increment(ref _criticalSectionDepth);
+            _criticalSectionSemaphore.Wait(millisecondsTimeout: 0);
```

Comment:
It's not really trying to obtain a lock in a traditional sense.  It's more like a flag that you can synchronously wait until it's false if that makes sense.  Callers to this method only care about asserting that the pipeline stopper will be halted until (at minimum) they call `Exit`.

Reply:
I think that's fine. The only need we have to _wait_ necessarily should be in `StopProcessing()`.

I don't think we have a need to force entrant calls to wait. Note that forcing entrant calls to wait for the semaphore to be available would cause a pattern like this to instantly deadlock and never complete:

```ps
& {
    cleanup {
        & { 
            cleanup { 
                # some code
            }
        }
    }
}
```
Or indeed any cmdlet calling any other cmdlet inside its cleanup block, potentially could cause a deadlock, which is _super_ no bueno I think. I'm unsure if this would be a concern for script cmdlets calling compiled cmdlets, but _definitely_ needs to be avoided if we want to support script cmdlets calling other script cmdlets inside their `cleanup{}` block.

Reply:
> It's not really trying to obtain a lock in a traditional sense

So with a different method name, and if the method were private, I think that would be fine.

But with internal methods called `EnterCriticalSection()` and `ExitCriticalSection()`, most people working here are going to read that as lock acquisition and release.

I'm very very reluctant to say we should add a complicated new concurrency mechanism deep within the PowerShell engine, especially because .NET already provides a number of nice ones (rich enough that I have yet to see a real use case for some of them). For example:

> It's more like a flag that you can synchronously wait until it's false if that makes sense

This is provided by [ManualResetEventSlim](https://docs.microsoft.com/en-us/dotnet/api/system.threading.manualreseteventslim?view=netcore-3.1).

Reply:
But the other big reason for this is that my experience with concurrency falls into one of three scenarios:

- I need an existing concurrency primitive
- I don't yet understand the concurrency problem I'm trying to solve (and think I need a new concurrency primitive, because I haven't yet seen the shape of the problem)
- I need a hack because the architecture of my problem is wrong but I can't change that right now

Basically, either an existing concurrency concept applies, or something is wrong on our side. Struggling to find a good way to manage our concurrency suggests the latter to me.

Reply:
Anyway, my hope is to be able to sit down at some point during the week and understand the concurrency demands here a bit better so that I can make a concrete proposal to you, rather than picking apart various usages. Hard to review concurrent code snippet by snippet.

Reply:
> > It's more like a flag that you can synchronously wait until it's false if that makes sense
> 
> This is provided by [ManualResetEventSlim](https://docs.microsoft.com/en-us/dotnet/api/system.threading.manualreseteventslim?view=netcore-3.1).

Yeah sorry I was over simplifying.  The part that `ManualResetEventSlim` doesn't do is re-entrance.